Source/WebCore/platform/CaretAnimation.h

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright (C) 2022 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+
+#include "Timer.h"
+#include "wtf/Forward.h"
+#include "wtf/Seconds.h"
+#include "wtf/StdLibExtras.h"
+#include <wtf/FastMalloc.h>
+#include <wtf/MonotonicTime.h>
+
+namespace WebCore {
+
+class CaretAnimation;
+
+class CaretAnimationClient {
+public:
+    virtual ~CaretAnimationClient() = default;
+
+    virtual void caretAnimationDidUpdate(CaretAnimation&) { }
+    virtual void caretAnimationWillStart(CaretAnimation&) { }
+    virtual void caretAnimationDidEnd(CaretAnimation&) { }
+};
+
+class CaretAnimation {
+    WTF_MAKE_FAST_ALLOCATED;
+public:
+    enum class Type {
+        Simple
+    };
+
+    CaretAnimation(Type animationType, CaretAnimationClient& client) : m_animationType(animationType), m_caretBlinkTimer(*this, &CaretAnimation::fired), m_client(client) { }
+
+    virtual ~CaretAnimation() = default;
+
+    Type type() const { return m_animationType; }
+
+    virtual void start(MonotonicTime currentTime, Seconds interval) = 0;
+
+    virtual void stop()
+    {
+        if (!m_isActive)
+            return;
+        didEnd();
+    }
+
+    virtual bool isActive() const { return m_isActive; }
+
+    virtual void serviceAnimation(MonotonicTime) = 0;
+
+    virtual String debugDescription() const = 0;
+
+    void fired() {
```

Comment:
Nit: { should be on new line.

Reply:
@rniwa This PR isn't meant for review, its just a draft PR for my own sake :)